### PR TITLE
Widemem - fixes to run trusty

### PR DESCRIPTION
**Pull request by priyasiddharth**

### CODE REVIEW

File: include/seahorn/Expr/Smt/Z3.hh [hunk 1]
```diff
@@ -482,6 +494,20 @@ public:
     Expr res = z3.toExpr(z3::ast(ctx, Z3_simplify_ex(ctx, ast, m_params)),
                          m_ast_to_expr);
     m_cache.insert({e, res});
+    LOG(
```
Comment (agurfinkel) [inline include/seahorn/Expr/Smt/Z3.hh:497]: the LOG messages do not belong here. They are strictly for bvopsem2

File: lib/seahorn/BvOpSem2.cc [hunk 2]
```diff
@@ -530,7 +530,9 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
       auto *val = CS.getArgument(2);
       Expr symVal = lookup(*val);
       if (symVal && arg0 && arg1) {
-        res = m_ctx.alu().Extract({symVal, val->getType()->getScalarSizeInBits()}, arg0->getZExtValue(), arg1->getZExtValue());
+        res =
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:533]: space issue

File: lib/seahorn/BvOpSem2.cc [hunk 3]
```diff
@@ -1765,6 +1767,13 @@ Bv2OpSemContext::Bv2OpSemContext(Bv2OpSem &sem, SymStore &values,
   zeroE = mkTerm<expr::mpz_class>(0UL, efac());
   oneE = mkTerm<expr::mpz_class>(1UL, efac());
 
+  m_z3.reset(new EZ3(efac()));
+  m_z3_simplifier.reset(new ZSimplifier<EZ3>(*m_z3));
+  auto &params = m_z3_simplifier->params();
+  params.set("ctrl_c", true);
+  if (SimplifyExpr) {
+    m_shouldSimplify = true;
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:1775]: `m_shouldSimplify = SimplifyExpr;`

File: lib/seahorn/BvOpSem2.cc [hunk 4]
```diff
@@ -1794,48 +1803,12 @@ Bv2OpSemContext::Bv2OpSemContext(SymStore &values, ExprVector &side,
...
-    if (strct::isStructVal(u)) {
-      llvm::SmallVector<Expr, 8> kids;
-      for (unsigned i = 0, sz = u->arity(); i < sz; ++i)
-        kids.push_back(m_z3_simplifier->simplify(u->arg(i)));
-      _u = strct::mk(kids);
-    } else {
-      _u = m_z3_simplifier->simplify(u);
-    }
-
-    LOG(
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:1823]: these log messages have to be moved back from Z3.hh to here

File: lib/seahorn/BvOpSem2Context.hh [hunk 5]
```diff
@@ -110,6 +112,11 @@ public:
   ~Bv2OpSemContext() override = default;
 
   EZ3 *getZ3() const { return m_z3.get(); }
+  std::shared_ptr<ZSimplifier<EZ3>> getSimplifier() {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2Context.hh:115]: Most of the time you would want to return `ZSimplifier<EZ3> &` 

What is the reason for sharing ownership with the caller?

File: lib/seahorn/BvOpSem2ExtraWideMemMgr.cc [hunk 6]
```diff
@@ -544,6 +547,11 @@ class ExtraWideMemManager : public OpSemMemManagerBase {
 
   Expr isDereferenceable(PtrTy p, Expr byteSz) {
     // size should be >= byteSz + offset
+    if (m_ctx.shouldSimplify()) {
+      ScopedStats _st_("opsem.simplify");
+      p = PtrTy(m_ctx.getSimplifier()->simplify(p.toExpr()));
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2ExtraWideMemMgr.cc:552]: does this simplification has a big effect? 

I'd imagine bigger effect on size and offset portion

**Comment by agurfinkel** [comment]
Needs a few tweaks before I can accept it.

File: include/seahorn/Expr/Smt/Z3.hh (hunk 1 above)
Comment (priyasiddharth) [reply include/seahorn/Expr/Smt/Z3.hh]: I have moved them back... can you please let me know why though?

File: lib/seahorn/BvOpSem2.cc (hunk 2 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:533]: This has been auto reformatted by the editor(vscode) using clang-format. If this is a issue with not following clang-format, I can fix it.

File: lib/seahorn/BvOpSem2.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: done!

File: lib/seahorn/BvOpSem2Context.hh (hunk 5 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2Context.hh]: Fixed... may i know the reason for making it a shared_ptr in the first place... who is an entity we want to share the simplifier with?

File: lib/seahorn/BvOpSem2.cc (hunk 2 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2.cc:533]: ok. maybe it just looks strange in the web interface

File: lib/seahorn/BvOpSem2ExtraWideMemMgr.cc (hunk 6 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2ExtraWideMemMgr.cc]: Good point! I fixed to only simplify offset, size sampling we don't seem to be spending much time in `simplify` but rather `solver::check_sat()`

File: lib/seahorn/BvOpSem2Context.hh (hunk 5 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2Context.hh]: Not really sure. Maybe there was some issue with copy constructor or something like that. We can try moving it to `std::unique_ptr` instead.

**Comment by priyasiddharth** [comment]
Ok to review... once if passes the presubmit checks.

File: include/seahorn/Expr/Smt/Z3.hh (hunk 1 above)
Comment (agurfinkel) [reply include/seahorn/Expr/Smt/Z3.hh]: the logs are specific for debugging bvopsem. They are of no use for other uses of the simplifier

**Pull request merged**